File: kotlin/src/main/kotlin/controllers/ProcesosController.kt
package controllers

import errors.ProcesoException
import models.Proceso
import repositories.ColaPrioritaria

/**
 * Esto no es obligatorio, solo lo uso para usar Mockito y puedeas ver una forma de testear
 * Lo ideal es solo hacer la cola
 * Aplico Inyección de Dependencias por constructor
 */
class ProcesosController(private val procesos: ColaPrioritaria) {

    fun push(proceso: Proceso) {
        procesos.push(proceso)
    }

    fun getById(id: Int): Proceso {
        return procesos.getById(id) ?: throw ProcesoException("No existe el proceso con id $id")
    }

    fun pop(): Proceso {
        return procesos.pop() ?: throw ProcesoException("No existe el procesos o la cola esta vacía")
    }

    fun isEmpty(): Boolean {
        return procesos.isEmpty()
    }

    fun getAll(): List<Proceso> {
        return procesos.getAll()
    }

}

File: typescript/src/controllers/ProcesosController.ts
import Proceso from '../models/Proceso'
import ColaPrioritaria from '../repositories/ColaPrioritaria'
import ProcesosException from '../errors/ProcesosException'

class ProcesosController {
  private readonly cola: ColaPrioritaria

  constructor(cola: ColaPrioritaria) {
    this.cola = cola
  }

  public push(proceso: Proceso): void {
    this.cola.push(proceso)
  }

  public getById(id: number): Proceso {
    const proceso = this.cola.getById(id)
    if (proceso != null) {
      return proceso
    }
    throw new ProcesosException(`No existe el proceso con id ${id}`)
  }

  public pop(): Proceso {
    const p = this.cola.pop()
    if (p != null) {
      return p
    }
    throw new ProcesosException('No existe el procesos o la cola esta vacía')
  }

  public isEmpty(): boolean {
    return this.cola.isEmpty()
  }

  public getAll(): Proceso[] {
    return this.cola.getAll()
  }
}

export default ProcesosController
